Avoid re-syncing LBs for ETP=local services

alexanderConstantinescu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->


/kind bug

#### What this PR does / why we need it:


This PR reduces the amount of LB re-configurations for ETP=local services by removing the sync completely for any node readiness changes, from the CCM's sync loop. The referenced issue explains the problem in greater detail, so readers are henced directed to that for what concerns the genesis of the problem. A lot has been discussed in this PR and a plan forward has been drafted. The following summarizes the discussion:

- Remove syncs w.r.t node state changes in the CCM in 1.25 (follow up PR, once this merges, is to be expected for remaining node conditions) 
- Maybe / possibly / hopefully backport that to the oldest supported Kube version also supporting ETP=local services
- Have the reported status of the LB HC configured for ETP=local services reflect the node state in 1.25 (also in the form of a follow up PR)
- Write KEP for dealing with transitioning node state for all service LBs. 
-  Kube >= after KEP : remove the sync logic for node state changes for all services and move them to the HC 


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Partially fixes #111539 for ETP=local services, specifically: reduces the amount of cloud API calls and service downtime caused by excessive re-configurations of all cluster LBs as an effect of transitioning node readiness state. 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Reduce the number of cloud API calls and service downtime caused by excessive re-configurations of cluster LBs with externalTrafficPolicy=Local when node readiness changes (https://github.com/kubernetes/kubernetes/issues/111539). The service controller (in cloud-controller-manager) will avoid resyncing nodes which are transitioning between `Ready` / `NotReady` (only for for ETP=Local Services). The LBs used for these services will solely rely on the health check probe defined by the `healthCheckNodePort` to determine if a particular node is to be used for traffic load balancing. 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig network
/sig cloud-provider

kinderyj commented:
a

leilajal commented:
/remove-sig api-machinery
/triage accepted

bowei commented:
cc: @bowei

bowei commented:
/assign @bowei -- I'm very interested in poking more at the lifecycle of the LB node

alexanderConstantinescu commented:
Some additional notes here below from the recent discussion in sig-net that just ended. Also, you'll be able to find the slides presented by me, here: https://docs.google.com/presentation/d/1EsK5FLS4s6E3Ok42XhRz643YtB9uZHn3wohWvcMyHag/edit#slide=id.p

My third slide present some ideas around the possibility of removing everything the CCM does w.r.t synchronizing nodes transitioning between `NotReady` <-> `Ready`.  

1. Is this PR - which eliminates the mentioned sync for ETP=local services, from the CCM's node sync loop
2. Is removing that sync completely for all services irrespective of ETP=local or not. 
a. Is removing that sync completely for a selected set of services, indicated by the user through the utilization of an annotation on the service object 
3. Is removing that sync completely for ETP=local services from the CCM, but replacing it by having kube-proxy dial the kubelet's read-only port when answering to the probe executed by the LB against the `HealthCheckNodePort`. I.e: kube-proxy would keep the current behavior (validating that the node in question is actually hosting endpoints for the service) and AND-ing that with the kubelet's reported state. That means that the LB would continue executing one probe against the `HealthCheckNodePort`, but kube-proxy would open a local connection against kubelet's read-only port on the node for every LB probe. There was some confusion surrounding this during the meeting, hence why I've chosen to detail this here. This would result in the same behavior as today, but be much more dynamic and efficient at handling transitioning node state - where instead of re-configuring the LB with a new set of nodes, the LB simply probes the node on the `HealthCheckNodePort` as it currently does, with the result also reflecting the kubelet/node readiness state. 
4. Is removing that sync completely for _all_ services from the CCM. Instead we define a standardised behavior for health checking nodes which are listed as backends for `LoadBalancer` type services, that all cloud provider implementations must adhere to. All cloud providers today implement this interface: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/cloud.go#L133-L162 , however: when the service is ETP != local they are free to specify custom probes on their LBs. What they do today was described at the bottom of the description to this PR. If we align the behavior for all cloud implementations: we could have them include a probe against the kubelet's read-only port - or do as in the ETP=local case and have the LB query kube-proxy which does multiple checks. In the end, LB's only really care to know if the node can service traffic and if the service `NodePort` on the node is reachable. All three clouds have chosen one of these two checks, not both. GCE's LB check is actually superfluous, since the CCM also checks the same condition (i.e: the kubelet state/node readiness).  

3. and 4. will require enhancement proposal and are beyond the scope of this PR, but we can discuss them here (or I can open a KEP for them and we can discuss them there...?)

This PR only concerns itself with finding a solution to either 1. and/or 2., simply because this currently is a bug for clusters at scale and, I believe, we should backport the fix to older releases. 

I am actually in favour of doing 2.a., meaning: modifying this PR slightly and removing any relation to ETP=local services and checking where endpoints are hosted. I would instead prefer we define an annotation (ex: `service.beta.kubernetes.io/lb-ignore-node-not-ready`) which a user can set on the service object to tell the CCM to ignore this service for node readiness changes. The implementation effort would be minimal and would solve both problems presented in my first slides. I, as a user of Kubernetes, should be allowed to disable the CCM's sync for node readiness state if I don't want my LBs to be re-configured as a cause by that. I own my nodes and LBs and should thus be allowed to decide what's best...if that results in the node actually going south as a consequence of my actions, so be it.    

If we can't agree on 2.a, then 1. (the current state of this PR) will at least help reduce the time taken to sync all LBs by the CCM on large clusters with a lot of services ETP=local, which is already an improvement from today.

alexanderConstantinescu commented:
/remove-sig api-machinery

danwinship commented:
/ok-to-test

alexanderConstantinescu commented:
/test pull-kubernetes-unit

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -762,25 +786,62 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
+		//   c) Ready -> NotReady
+		//   d) NotReady -> Ready
+		// We only want to possibly __not__ sync on c) and d).
+		if len(oldNodes) == len(newNodes) {
```
I feel like I am missing something.  Why do we care if the lengths match?  Is it possible that I added a node an removed a different node fast enough that this captures both?

It feels like this could be easier to follow, am I missing some finer point?

```
// For services of ETP=local: we do not want to sync nodes transitioning from NotReady <-> Ready
// if these hosts are not actually hosting any endpoints for the service. Doing that leads to
// unnecessary updates for LBs which will never send traffic to the nodes in question in any case.
if svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal {
	needsSync := func() bool {
		// We have four cases to consider here:
		//   a) Node ADD
		//   b) Node DEL
		//   c) Node UPDATE Ready -> NotReady
		//   d) Node UPDATE NotReady -> Ready
		// In case of (c) or (d) we can skip sync for ETP=Local Service
		oldNodeNames := nodeNames(oldNodes)
		newNodeNames := nodeNames(newNodes)
		if !oldNodeNames.Equal(newNodeNames) {
			// The set of nodes has changed - needs sync.
			return true
		}
		// The set of nodes didn't change, but maybe readiness did.
		oldReadyNodes := filterWithPredicate(oldNodes, s.getNodeConditionPredicate(true))
		newReadyNodes := filterWithPredicate(newNodes, s.getNodeConditionPredicate(true))
		oldReadyNodeNames := nodeNames(oldReadyNodes)
		newReadyNodeNames := nodeNames(newReadyNodes)

		// Figure out which nodes changed readiness.
		changedNodeNames := oldReadyNodeNames.Difference(newReadyNodeNames)
		changedNodeNames = changedNodeNames.Union(newReadyNodeNames.Difference(oldReadyNodeNames))
		if len(changedNodeNames) == 0 {
			// Something else changed - needs sync.
			return true
		}

		if s.serviceHasEndpointsOnNodes(svc, changedNodeNames) {
			// This service has at least 1 endpoint on a changed
			// node - needs sync.
				return true
		}
		// It seems safe to skip this synv for this service.
		return false
	}
	if !needsSync() {
		return false
	}
}
```

thockin replied:
I put the "Something else changed - needs sync" block in as the sort of "default" case - something I don't understand happened and it seems safer to sync.  We should probably log something useful here, since it should not happen.

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -762,25 +786,62 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
+		//   d) NotReady -> Ready
+		// We only want to possibly __not__ sync on c) and d).
+		if len(oldNodes) == len(newNodes) {
+			if newNotReadyNodes.Len() > 0 && !s.serviceHasEndpointsOnNodes(svc, newNotReadyNodes) {
```
I think, as written, this will skip sync on services that need sync.  If one node became unready with no endpoints and one became ready (with endpoints) the first case would hit and it would return without evaluating the second case?

alexanderConstantinescu replied:
Yes, quite possibly. I will try to extend my tests with a more exhaustive list of corner cases, one of which should include this.

thockin reviewed (commented):
I didn't look at the tests yet.

I admit I am anxious about this - I need to think through the possible implications.  I don't like this controller looking at endpoints - I'd rather we had a cleaner lifecycle, like your proposal alternatives lead to.  That will take more time, though, and I ACK that this problem is real.

thockin commented:
Thinking more about this: For node-centric LB setups, there are 3 states a Node can be in wrt LB backend sets:

1) The node is not in the LB set(s)
2) The node is in the LB set(s) but the LB's healthcheck to that node is failing
3) The node is in the LB set(s) and the LB's healthcheck to that node is passing

Today we consider a variety of distinct signals. which add up to mild chaos:
* Is the Node Ready (as per Conditions)?
* Is the LB's healthcheck to that node passing?
* Does the node have the "exclude" label?
* Does the node have to-be-deleted taint?
* Is the node Unschedulable?

With the caveat that different implementations have different healthcheck capabilities, I think that we should decide and document the ideal mapping of these individual inputs to the 3 primary states, with an emphasis on minimizing API calls to the management plane while delivering the most-correct behavior possible (and to be clear I think we are NOT most-correct today).

I posit that the ONLY reasons to move a node out of the LB set are "administrative action" - specifically someone set the `node.kubernetes.io/exclude-from-external-load-balancers` label.  Otherwise, once it is in the LB set, we should leave it there and rely on healthchecks.

Why?  Two reasons.

1) As evidenced by this PR, adding/removing from the set is expensive and should be avoided when possible.
2) In some clouds there is a distinction between "in the LB set but failing HC" and "not in the LB set".   In the case of the former, existing flows are preserved but new connections will avoid the node.  In the case of the latter, existing flows are broken.  Keeping the LB in the set means we can at least TRY to service existing connections.

![Node_LB_states](https://user-images.githubusercontent.com/5595220/178065317-3629ecb6-4816-49c2-8b21-74c06b062661.png)

How?  The healthchecks we set up for LBs need to be smarter.  We can build this into kube-proxy (and every other kube-proxy equivalent) or we can build a different mechanism - TBD.

This is obviously not helpful RIGHT NOW to this PR, but directionally - thoughts?

bowei commented:
@swetharepakula -- (see above diagram) -- this likely intersects with the endpoints lifecycle work

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -668,7 +690,7 @@ func nodeSlicesEqualForLB(x, y []*v1.Node) bool {
 }
 
-func (s *Controller) getNodeConditionPredicate() NodeConditionPredicate {
+func (s *Controller) getNodeConditionPredicate(excludeNotReadyNodes bool) NodeConditionPredicate {
```
> This is already a kind of double-negative, which makes my brain spin.

Looking at this PR ~ two months later: it makes mine spin too. I will apply your proposed change.

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -762,25 +786,62 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
 
 // nodeSyncService syncs the nodes for one load balancer type service
-func (s *Controller) nodeSyncService(svc *v1.Service) bool {
+func (s *Controller) nodeSyncService(svc *v1.Service, oldNodes, newNodes []*v1.Node) bool {
```
I changed the signature to this because I wanted to have a synchronized retrieval of `nodes`, and update of `lastSyncedNodes` in `updateLoadBalancerHosts`. `nodeSyncService` is parallelized per `workers` and I didn't want to run the risk of having different workers executing `nodeSyncService` with a different set of nodes - which might happen if we list nodes internally. 

I can comment on the assumptions.

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -762,25 +786,62 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
 }
 
 // nodeSyncService syncs the nodes for one load balancer type service
```
Will comment

alexanderConstantinescu commented:
> This is obviously not helpful RIGHT NOW to this PR, but directionally - thoughts?

I completely agree. And is also why I gravitated towards an enhancement proposal for this with an initial first draft of said proposal being: modifying the health check probes to what I mentioned in point [3. and 4. here](https://github.com/kubernetes/kubernetes/pull/109706#issuecomment-1112737473) . This should address the issue of reacting more efficiently to node state changes without also needing to re-configuring the LB on the cloud provider's side and disrupting ingress/egress traffic. 

I can start writing up the proposal, if we agree that we need it.   

I also want to emphasize that the referenced issue just above (#102367) will not get fixed by this PR if they don't have a large amount of ETP=local services, only mentioning this https://github.com/kubernetes/kubernetes/issues/102367#issuecomment-1179400517 mentions "I think the 'real fix' is related", and I want to set the expectation to anyone reading that, that: it is if you have a lot of "ETP=local" services, otherwise this won't fix anything.

alexanderConstantinescu commented:
/remove-sig api-machinery
/remove-sig testing

Updated the PR, taking into account the code review thus far.

thockin commented:
> I can start writing up the proposal, if we agree that we need it.

Yeah, I think your #4 is close to what I have in mind, though I don't think it's necessarily kubelet's healthz that LBs should probe - this is a place to explore options and tradeoffs.

Worth a KEP, I think.

thockin reviewed (commented):
Something smells wrong, but it took me a moment to fput my finger on it - suppose this:

Nodes A, B, C are ready.
Service S with ETP=local running on nodes A, B.
LB set for S is {A, B, C}.
Node A goes Unready.
We remove A from the LB set, now {B, C}.
Pod on A is removed.
Node A goes Ready.
We ignore it because no endpoints.
Start a new S pod on node A.

The LBSet is {B, C}.  C fails healthcheck so the effective set is {B}, despite a pod on A.  No Service or Node event will trigger to cause the reconcile.

If you want to get pathological, repeat the above with B:

Service S now has pods on nodes A, B.
LB set for S is {B, C}.
Node B goes Unready.
We remove B from the LB set, now {C}.
Pod on B is removed.
Node B goes Ready.
We ignore it because no endpoints.
Start a new S pod on node B.

The LBSet is {C} - the service is "down" despite 2 running pods.

Am I missing how this doesn't happen?

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -58,6 +63,10 @@ const (
 	// https://github.com/kubernetes/autoscaler/blob/e80ab518340f88f364fe3ef063f8303755125971/cluster-autoscaler/utils/deletetaint/delete.go#L36
 	ToBeDeletedTaint = "ToBeDeletedByClusterAutoscaler"
+	// includeOnlyReadyNodes enables the node condition predicate and includes only all Ready nodes.
+	includeOnlyReadyNodes nodeInclusionMode = 1 << iota
```
I opted for not using a bitmask

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -761,26 +792,86 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
+			if len(changedNodeNames) == 0 {
+				// Something else changed, but LBs are not concerned by the
+				// change, skip sync.
+				return false
```
I've re-worked the code quite a bit and added the part which performs this validation in its own separate (last) commit. It might need some re-working, but I don't have any good ideas for the time being.

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -761,26 +792,86 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
+// The return value indicates if we should retry. Hence, this functions returns false
+// if we've updated load balancers and finished doing it successfully, or didn't try to
+// at all because there's no need. This function returns true if we tried to update load
+// balancers and failed, indicating to the caller that we should try again.
```
Re-worked the code and given its recent state, I don't think this is needed.

thockin commented:
So this is now pretty complex.  I assert that we don't REALLY want any of this - we really want something like the diagram above, which should totally solve the problem and not need any of this complexity (so this ends up being reverted).

I'm also worried about possible other implementations, like GCP's ILB subsetting, not being kept in sync.

What if we, instead, actually tackled the real issue?  I strarted this a while back, though I feel like this could use some cleanup:  https://github.com/thockin/kubernetes/commit/9dd360185becfb972c7777f7c83b5246909c3df9

Specifically, what if we :

1) Change service controller to NOT EXCLUDE unready, unschedulable, and tained nodes.  LBs would be given all nodes that are not excluded by the specific exclsuion label.

2) Change kube-proxy to watch its own Node and make healthz fail if the node is Unready, Unschedulable, or ToBeDeleted tainted.

Results: 

* LB impls that use that healthz will work better.
 
* LB impls that don't use that healthz will see nodes that they didn't see before.  I argue that this may be acceptable because:
  - if the node is legit dead, the LB needs some HC anyway which will cover this
  - if the node is being drained (unschedulable) or being deleted (tainted) then the eventual node removal is just as abrupt as removing it from the LB set (no opportunity to drain), so "not worse than before".

* In the face of version skew (older kube-proxy, newer control-plane), LB impls will see nodes that they didn't see before.   See previous point.

* We could make a new URL (e.g. /readyz) but then we need to wait until the oldest supported kube-proxy supports this, which means introducing the kube-proxy change, waiting a year, then implementing the controller half.  I don't think that is warranted, given the above.

* We need to update alternative implementations (e.g. Cilium, Antrea(?), etc) to also HC-fail in these node situations.

It's a deeper change but it's actually one we want to keep.  What do you all think?  Am I trivializing the risks?

alexanderConstantinescu commented:
> So this is now pretty complex 

Yes, I agree...but you should see the work-arounds to this problem that users are thinking about deploying.

 > I assert that we don't REALLY want any of this - we really want something like the diagram above, which should totally solve the problem and not need any of this complexity (so this ends up being reverted).

I 100 % agree that we want that diagram. The reason as to why I never started doing that was because I supposed that an enhancement proposal would be required for such a change. Wouldn't doing such a thing diverge the implementation / expectation that is already set by different Kubernetes service proxies? I thought that's the point of the enhancement proposal and why we do them to begin with: to align and coordinate different implementations/vendors, so that users don't suffer.

If that is not the case I could obviously modify this PR (or build on your: https://github.com/thockin/kubernetes/commit/9dd360185becfb972c7777f7c83b5246909c3df9) to implement that mechanism instead. 

One thing though, in your diagram you noted: 

> It's really about whether the on-node proxy mechanism is up-to-date and ready to handle traffic 

but isn't that already the case for ETP=local services through the usage of `healthCheckNodePort` probe? That probe from the LB to the proxy evaluates precisely if the proxy on each node can respond, and moreover: even execute logic which determines if the service has an endpoint on the node. If we want to implement that diagram for ETP=local services, we might as well just remove the current CCM logic for ETP=local services altogether. Just configure LBs with the full set of nodes (minus the ones with the exclusion label) and avoid re-configuring LBs if nodes experience readiness changes.   

My hope and original goal _with this PR_ was to back-port this PR to older releases once we have something we agree upon...but given the turn this discussion is taking, maybe that was a wild dream. But in that case I would even say: let's not do anything for now and instead write up an enhancement proposal with a long term vision for this that includes all service proxies/CNI plugins and cloud providers. 

> Am I trivializing the risks? 

I can't speak for that fully since I lack visibility into the LB part on the cloud provider's side to tell, but I don't see why. The current problem (re-configuring all LBs for every node change in a cluster) in the CCM is a rather important issue. Implementing the mentioned behaviour would keep the functional behaviour aligned with the current, but be more efficient. It's more of having all stakeholders aligned on the expectations.

thockin commented:
> The reason as to why I never started doing that was because I supposed that an enhancement proposal would be required for such a change

It would, but that doesn't mean "Oh no, it will take 18 months".  It just means we document the changes and tradeoffs BEFORE we make them in code.  It will require some outreach and "doing the legwork", of course but it's not fatal.  The main drawback is that the KEP window for 1.25 is already closed, so it could not  be launched until 1.26, really.

> I thought that's the point of the enhancement proposal and why we do them to begin with: to align and coordinate different implementations/vendors, so that users don't suffer.

I *think* we can do it in a way that is "not worse than today", which alleviates a lot of the pain.

> One thing though, in your diagram you noted:
>
>> It's really about whether the on-node proxy mechanism is up-to-date and ready to handle traffic
>
> but isn't that already the case for ETP=local services through the usage of healthCheckNodePort probe? That probe from the LB to the proxy evaluates precisely if the proxy on each node can respond, and moreover: even execute logic which determines if the service has an endpoint on the node. 

Sort of.  I took a quick peek at this and it seems to only check if there are any local endpoints.

We have one handler (pkg/proxy/healthcheck/proxier_health.go) which actually considers whether the proxy is healthy (though I am skeptical that it is really that robust, in looking at it again).

For ETP=Local we have another handler (pkg/proxy/healthcheck/service_health.go) which only considers whether therer are local endpoints.

That 2nd one should probably be the AND of both.

I'm saying that both of these should ALSO consider `unschedulable` and `ToBeDeleted` taints.  But even if they don't it's not a huge deal (I think - this is where we need some extra eyeballs).

> If we want to implement that diagram for ETP=local services, we might as well just remove the current CCM logic for ETP=local services altogether. Just configure LBs with the full set of nodes (minus the ones with the exclusion label) and avoid re-configuring LBs if nodes experience readiness changes.

Now THAT is a good idea.  I think we should *probably* do this for all Services, but doing it JUST for ETP=Local could be the first move.  That solves your problem without adding all the endpoints logic and moves us closer to where we want to be.

At the same time, we could change the endpoints to be more robust, and get that rolling out.  Since that is just moving equivalent logic from the service controller to the HC, I don't think we need a KEP.  The service logic remains in place, except ETP=Local.

The only real behavioral change, other than avoiding LB set syncs, is this:  In the face of version skew (an old kube-proxy with a new control-plane) an ETP=Local LB service will send traffic to nodes which are Unready or Unschedulable or tainted, but _only if_ those nodes have at least one endpoint for the Service.    And I still posit that this is OK - if the node goes hard-down, we have the same abrupt termination situation.  If the node is cleanly drained, those endpoints will go away.

1.25 could include the improved HCs and ETP=Local change.  During 1.26 we could KEP and discuss the equivalent for everyone else.

> My hope and original goal with this PR was to back-port this PR to older releases

If we just disable Unready-sync for ETP=Local, backporting is POSSIBLE, but we'll need to really justify it, That may be tough.

alexanderConstantinescu commented:
> The main drawback is that the KEP window for 1.25 is already closed, so it could not be launched until 1.26, really.

OK, I will need to focus on some other tasks for a couple of days, but I will get the enhancement proposal going ASAP, as to try to meet the 1.26 deadlines. 

> Now THAT is a good idea. I think we should probably do this for all Services, but doing it JUST for ETP=Local could be the first move. That solves your problem without adding all the endpoints logic and moves us closer to where we want to be.

Great! I will update this PR with that logic. My misunderstanding of what get placed in the "this change really needs a KEP before we can do anything" bucket was flawed, otherwise I would have done that from the get-go. 

> If we just disable Unready-sync for ETP=Local, backporting is POSSIBLE, but we'll need to really justify it, That may be tough.

Who do we need to justify it to? Maybe we can start poking people on this?

I think removing the CCM's LB set sync for ETP=local is an improvement from today, even without any change to any HC. Large clusters experience more entropy, entropy leads to more LB re-configuration, LB re-configuration leads to impacted SLAs. The changes leading to increased entropy might not have any relation to the service answering to that SLA...but currently it might/will.

alexanderConstantinescu commented:
I think this discussion really boils down to: "what does a node going NotReady really say about the networking state of the node?" - I feel like we don't know the answer. And maybe that should be the first question to work out on the enhancement proposal 

We had similar discussions/questions in OpenShift concerning what actions surrounding networking components should take for nodes that experience transient state to the negative - should we forward traffic to them or start ignoring them? I think we just based/copied the assumptions made by Kube...I was never sure that was correct, and I still am not. Like, in terms of causality: is there really any networking failure that will transition the node into `NotReady`, besides the classic and evident "CNI plugin hasn't written it's config file to disk". But that's more like a day 0 issue which means we probably don't have any endpoints on the node in any case, because nothing could be started. Then again, the kubelet going wonky might be because there's disk/CPU/memory pressure on the node, which might be impacting the kernel too, which might be impacting networking. But this is very speculative. I don't think there's ever been a clear and definitive networking test perform by the kubelet, which would place the node in the failed state and justify a load balancer/router/proxy/other node to stop sending traffic to the node in question - especially if the load balancer/router/proxy/other node already uses a dedicated probe to test if it can communicate with the pod/kube-proxy/node and _that passes_. So either we think of such a test that the kubelet could perform as to have NotReady actually mean that "this node should stop receiving traffic", or we agree that no such test can be designed because "networking is hard" and we just rely on HC and stop thinking about the node's readiness state for what concerns routing network traffic. 

I am saying this because even ETP=cluster services enable a health check from the LB to the nodes (on the major public cloud providers), as I mentioned in my PR description:

> GCE: probes [10256](https://github.com/kubernetes/kubernetes/blob/master/pkg/cluster/ports/ports.go#L39)
> AWS: if ELB: probes the first NodePort defined on the service spec.
> Azure: probes all NodePort defined on the service spec. 

So maybe we should remove the CCMs LB set syncs, for all services without any change to any HC? That's maybe asking too much, or asking for trouble....

alexanderConstantinescu commented:
About backporting: would we want to backport the complicated current fix which keeps the behaviour almost aligned with current, i.e: disable unready-sync for nodes with no endpoints? Or just disable unready-sync altogether, knowing that we won't replace it in the HC? 

I am thinking:

Kube < 1.25 - complicated-but-fully-aligned-with-current-behavior-fix in the CCM for ETP=local
Kube = 1.25 - simple fix (disable unready+taint-sync) for ETP=local in the CCM and place them in the ETP=local HC
Kube >= 1.26 (or whenever the KEP gets in) - disable all checks in the CCM and place them in the HC corresponding to each service type

alexanderConstantinescu commented:
OK, so what have I done you ask?

1. First commit contains the simplified logic for syncing unready nodes in the service controller for ETP=local (I kept this split with the idea of backporting this)
2. Expands that for the taint
3. Moves the taint and readiness checks to the HC for ETP=local services 
4. Includes the proxy healthz result in the HC for ETP=local services

This should capture:

 > We have one handler (pkg/proxy/healthcheck/proxier_health.go) which actually considers whether the proxy is healthy (though I am skeptical that it is really that robust, in looking at it again).
>
> For ETP=Local we have another handler (pkg/proxy/healthcheck/service_health.go) which only considers whether therer are local endpoints.
>
> That 2nd one should probably be the AND of both.
>
> I'm saying that both of these should ALSO consider unschedulable and ToBeDeleted taints. But even if they don't it's not a huge deal (I think - this is where we need some extra eyeballs).

This is a first version obviously and head start for this work. In particular the 4 commit, which is baking in the healthz server in the "service HC server" looks.....well, weird. But it's late at night for me and I thought it's that or just dialing the healthz server on localhost instead.

alexanderConstantinescu commented:
FWIW: I figured out a way to _vastly_ simplify the original PR, see: https://github.com/kubernetes/kubernetes/commit/0877f331cc33d93dae603a56da938153b418f859 

It dawned on me that `updateLoadBalancerHosts` actually keeps track of the `servicesToRetry` and re-syncs them if there are any. We hence don't need to keep track of all configured nodes per service, but instead only update the list of `lastSyncedNodes`, if we tried to sync and that went well. If that is not the case `servicesToRetry` will be tried again with the same set of nodes.

I don't know if this fact makes the "complicated" (and more in-line with today) version more likely to win points for "backportability"....but placing this here, so that people following along are aware.

thockin commented:
I don't understand https://github.com/kubernetes/kubernetes/commit/0877f331cc33d93dae603a56da938153b418f859 - I am missing something.  If any service Successfully syncs, then the lastSynced gets updated - how does that prevent the drift?

I'll look at the main commits in this PR now.

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -765,26 +789,58 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
+			// The set of nodes has changed - needs sync.
+			return true
+		}
+		// We are processing the same set of nodes, some might experience
```
The filtering logic in this file is so spread around that I am having a really hard time convincing myself that we can't end up here when something else changed.

I'm looking for a way to make it clearer what's happening as the control moves around and we list and re-list and re-re-list with different criteria.

Let's describe the final end-state:

* We sync all nodes (except those labelled "exclude-from-lb") to all LB services.
* Stat of and changes to Node Unschedulable, Taints, readiness are all irrelevant

Right?  And the intermediate state is:

* We sync all nodes (except those labelled "exclude-from-lb") to ETP=Local LB services.
* We sync a subset of nodes (excluding Unschedulable, Taints, and unready) to other LB service
* Changes to Node Unschedulable, Taints, readiness are relvant to ETP!=Local

So can we set the code up to reflect this?  Like, just assume that every list of nodes includes ALL nodes down-filter only at the end, when we look at each Service?

alexanderConstantinescu replied:
Let me see if I can summarize the logic as I've interpret it. Like you've said we react to node changes as:

> A Node was added
> A Node was deleted
> A Node was updated AND shouldSyncNode
>
> shouldSyncNode() is defined as any of:
>
> Node labels changed
> Node unschedulable changed
> Node ready changed

However `getNodeConditionPredicate` will perform a slightly different check, specifically looking at:

- Does the node has the "exclude label"
- Does it have the `ToBeDeletedTaint` key on its taints?
- Is it Ready or not 

Hence, as an example, we enqueue changes to `.Spec.Unschedulable`: but we don't consider it when actually syncing state. So even if it did change between two syncs, `getNodeConditionPredicate` doesn't check that attribute, so the set of nodes remain unchanged. What nodes get synced is the result of `getNodeConditionPredicate` only. 

I am reflecting on how big of an issue it is to not actually sync changes to `.Spec.Unschedulable`, given that we do sync changes to `.Spec.Taints` and specifically the: `ToBeDeletedTaint`. Maybe these go hand-in-hand? Doubt it...

> So can we set the code up to reflect this?

I am thinking about ways to make this better without changing too much.

thockin replied:
I'm OK with changing the code if it makes it simpler.  This code feel like it has evolved incorrectly - that we have different sets of conditions on different paths is a bit of a smell...

thockin reviewed (commented):
Sorry this is a little scattered - I ran out opf time to think more on it today, but I am struggling to track the logic.  It feels like it should be simpler, but I don't have time to which up a change to convince myself just now.

thockin reviewed (commented):
Also, I think we should do this as 3 PRs:

1) Less syncs for ETP Local
- don't sync when Ready changes
- don't resync when other things change

2) kube-proxy HC fail when Unschedulable, !Ready, etc.

3) Less syncs for all services

2 and 3 would be covered by the KEP and we'll need to vet 1 with some smart people to see if they can poke holes in the rationale for "it's safe".

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -251,7 +264,7 @@ func (s *Controller) Run(ctx context.Context, workers int) {
 	s.nodeSyncLock.Lock()
 	defer s.nodeSyncLock.Unlock()
-	newHosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
+	newHosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate(includeOnlyReadyNodes))
```
Taking a step back and looking at this function's goal, which I'd say was: determine if we should sync all services or only the services that might have error'd in sync N-1. `knownHosts` is used to set `needFullSync` to true, which is read in `needFullSyncAndUnmark` to determine if we should sync `servicesToUpdate` or "all services". 

Given that, I think this is right. Imagine this:

Sync 1: node1 (Ready), node2 (Ready), node3 (Ready) - newHosts: node1, node2, node3
Sync 2: node1 (Ready), node2 (Ready), node3 (Ready) - newHosts: node1, node2, node3
newHosts == knownHosts - only sync the which we've not been able to sync (`servicesToUpdate`)
Sync 3: node1 (NotReady), node2 (Ready), node3 (Ready) - newHosts: node2, node3
newHosts != knownHosts - sync all services
Sync 4: node1 (NotReady), node2 (Ready), node3 (excluded) - newHosts: node2
newHosts != knownHosts - sync all services
Sync 5: node1 (Ready), node2 (Ready), node3 (excluded) - newHosts: node1, node2
newHosts != knownHosts - sync all services
Sync 6: node1 (Ready), node2 (Ready), node3 (excluded) - newHosts: node1, node2
newHosts == knownHosts - only sync the which we've not been able to sync (`servicesToUpdate`)

I think the readability and logic could be improved across this function. But for what concerns this: it seems right.

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -690,12 +712,14 @@ func (s *Controller) getNodeConditionPredicate() NodeConditionPredicate {
-				klog.V(4).Infof("Ignoring node %v with %v condition status %v", node.Name, cond.Type, cond.Status)
-				return false
+		if include == includeOnlyReadyNodes {
+			for _, cond := range node.Status.Conditions {
```
Yes, will fix

alexanderConstantinescu committed:
De-duping node "update enqueuing"/sync predicates

alexanderConstantinescu committed:
Avoid re-syncing LBs for ETP=local svc

This removes the CCM's LB sync for nodes which are experiencing a transitioning
readiness state only.

alexanderConstantinescu commented:
> Also, I think we should do this as 3 PRs:
> 
> 1. Less syncs for ETP Local

Done. 

I've update the PR with a more readable and unified way of listing nodes based on predicates. I think this makes the node updates easier to read. I think that addresses your comments, in particular this one: https://github.com/kubernetes/kubernetes/pull/109706#discussion_r923896930

> So can we set the code up to reflect this? Like, just assume that every list of nodes includes ALL nodes down-filter only at the end, when we look at each Service?

Also:  based on the fact that we've unified `shouldSyncNode` with the predicates for listing nodes in the node sync loop: we now have an opportunity to clean up the service controller for what concerns syncing nodes. Specifically I wrote two additional commits which cleans up the node sync logic (https://github.com/kubernetes/kubernetes/commit/641dd23227ce053884136e832d33d300f8d7d4be) and also moves to using the workqueue for node syncs (https://github.com/kubernetes/kubernetes/commit/6fa06e94e2917adf862395987f0ce2f770715faf). I think these two commits make sense at this point, but I can file a new PR for them once we finish with this one.

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -443,7 +443,7 @@ func (s *Controller) syncLoadBalancerIfNeeded(ctx context.Context, service *v1.S
 }
 
 func (s *Controller) ensureLoadBalancer(ctx context.Context, service *v1.Service) (*v1.LoadBalancerStatus, error) {
```
super nit: the `s` should probably be `c` all over this file :)

alexanderConstantinescu replied:
I will fix this across the file, as a separate commit.

alexanderConstantinescu replied:
On second thought, it's quite a scattered nit in this file. I will open a seperate PR for that as to reduce the noise on this one.

thockin reviewed (commented):
These comments are just the first commit.  Overall I like the cleanup a lot, and I THINK it's mostly correct, but it's not really easy to prove, is it?  The old code was twisty.

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -941,6 +969,13 @@ var (
 
+func getNodePredicatesForService(service *v1.Service) []NodeConditionPredicate {
+	if service.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal {
+		return []NodeConditionPredicate{nodeIncludedPredicate, nodeSchedulablePredicate, nodeUnTaintedPredicate}
```
make another var `etpLocalNodePredicates` or move `allNodePredicates` into this function

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -559,6 +575,196 @@ func TestUpdateNodesInExternalLoadBalancer(t *testing.T) {
+	}
+
+	services := []*v1.Service{
+		newETPLocalService("s0", "777", v1.ServiceTypeLoadBalancer),
```
P{utting UID here feels like noise - do we need it somewhere?

alexanderConstantinescu replied:
No, it's a relic of copy-pasted code. I am removing it

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -559,6 +575,196 @@ func TestUpdateNodesInExternalLoadBalancer(t *testing.T) {
+	}
+
+	services := []*v1.Service{
+		newETPLocalService("s0", "777", v1.ServiceTypeLoadBalancer),
```
Do we really need 4 of these?  I don't see what this is testing over and abive 2 of them?

alexanderConstantinescu replied:
Yes, it's another relic of the previous versions of this PR when I was testing endpoints placements per ETP=local service. Given that we decided not to do that: let me do three services, two ETP=local and one ETP!=local.

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -559,6 +575,196 @@ func TestUpdateNodesInExternalLoadBalancer(t *testing.T) {
+				},
+			},
+			expectedUpdateCalls: []fakecloud.UpdateBalancerCall{
+				{Service: newETPLocalService("s0", "777", v1.ServiceTypeLoadBalancer), Hosts: []*v1.Node{node1, node2, node3}},
```
I dislike that this re-creates "s0", "s1", etc with args that have to match the above.  Can we just reference them directly?

thockin reviewed (commented):
2nd commit

thockin reviewed (commented):
Let's make the 3rd commit a new PR, so we can explain why it is safe without all the other noise?

thockin commented:
Before we commit, we should make sure the PR comment and relnote are good.

alexanderConstantinescu commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -1684,6 +1681,64 @@ func Test_shouldSyncNode(t *testing.T) {
 			},
```
I am adding these test cases now. Thanks for writing them down!

alexanderConstantinescu commented:
> Overall I like the cleanup a lot, and I THINK it's mostly correct, but it's not really easy to prove, is it? 

True. I suspect Kubernetes doesn't have a lot of disruption E2E tests for LoadBalancer type services (i.e: triggering nodes to fail and inspecting the outcome on the configured load balancers), which makes the validation a bit difficult. Does GKE have a more extended test suite for this? 

> Let's make the 3rd commit a new PR, so we can explain why it is safe without all the other noise? 

Will do.

> Before we commit, we should make sure the PR comment and relnote are good.

OK, I will update that now. 

The state of the PR now is:

- We cleanup the predicate evaluation for node updates and syncs 
- We don't sync node readiness changes for ETP=local services

I will open a separate PR for  https://github.com/kubernetes/kubernetes/pull/109706#discussion_r931619114 and an additional one for the previous third commit, as mentioned above.

BenTheElder commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -772,24 +748,29 @@ func (c *Controller) nodeSyncInternal(ctx context.Context, workers int) {
+// there's no need. This function returns true if we tried to update load balancers and
+// failed, indicating to the caller that we should try again.
+func (c *Controller) nodeSyncService(svc *v1.Service, oldNodes, newNodes []*v1.Node) bool {
+	retSuccess := false
+	retNeedRetry := true
```
it looks like these are intended to be named constants?
```suggestion
	const retSuccess = false
	const retNeedRetry = true
```

BenTheElder replied:
Though I would say the more usual approach would be to let the method documentation make it clear what true and false mean.

alexanderConstantinescu replied:
Would you say the current state of that doc bloc does a good job at making it clear what true and false mean? In that case, I could just remove this in a follow up PR.

thockin replied:
It's a weird signature (false usually means fail) so I like having the const values.

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -672,58 +672,13 @@ func nodeSlicesEqualForLB(x, y []*v1.Node) bool {
-		if condition.Type != v1.NodeReady {
-			continue
+	for _, p := range allNodePredicates {
+		if p(oldNode) != p(newNode) {
```
> I did that, but it wasn't in-line with the current behaviour.

OK, that's fair.  The current behavior is...pointless but not changing it is good.

thockin commented on staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go:
```diff
@@ -559,6 +575,193 @@ func TestUpdateNodesInExternalLoadBalancer(t *testing.T) {
 }
 
+func TestNodeChangesForExternalTrafficPolicyLocalServices(t *testing.T) {
+	node1 := &v1.Node{ObjectMeta: metav1.ObjectMeta{Name: "node0"}, Status: v1.NodeStatus{Conditions: []v1.NodeCondition{{Type: v1.NodeReady, Status: v1.ConditionTrue}}}}
```
minor but...the var says node1, the name is node0, which is super confusing with the next one - the name is node1 but the var is node2.  Can we just use the same number in both places?  I don't care if it is not 0-based :)

alexanderConstantinescu replied:
OK, I will make sure to fix this in the PR coming for the 3rd commit. This is recurrent across this test file, so maybe I'll add a dedicated PR for cleaning that up.

thockin replied:
SGTM

thockin commented:
Followup needed: 
* PR for 3rd commit
* KEP to change HC behavior and reduce all syncs - once in the set, only the exclusion label matters

Anything else?

I'm really counting on your follow-through on these - I'd hate it to get stuck half way. 

/lgtm
/approve

alexanderConstantinescu commented:
> Followup needed:
> 
> * PR for 3rd commit
> * KEP to change HC behavior and reduce all syncs - once in the set, only the exclusion label matters
> 
> Anything else?
 
Yes, maybe I drew wrong conclusions from our discussions on this PR, but I thought we also wanted to start moving the readiness logic to kube-proxy's HC for ETP=local, as to get a head start on the version skew?   

Specifically you mentioned here: https://github.com/kubernetes/kubernetes/pull/109706#issuecomment-1183782705:

> At the same time, we could change the endpoints to be more robust, and get that rolling out. Since that is just moving equivalent logic from the service controller to the HC, I don't think we need a KEP. The service logic remains in place, except ETP=Local.

> 1.25 could include the improved HCs and ETP=Local change. During 1.26 we could KEP and discuss the equivalent for everyone else.

For what concerns:

> I'm really counting on your follow-through on these - I'd hate it to get stuck half way.

Will do, I am ready to submit PRs for the 3rd commit and kube-proxy's ETP=local HC once this gets in. I will have the KEP up in a week.

thockin commented:
With the code freeze in effect, I am not sure we can get the HC changes done.  I wanted to break it out so that it didn't drag this main PR down with it :)

This PR was approved before freeze and is not part of a KEP (limited impact) - do we need an exception for it @cici37 ?

cici37 commented:
Thanks for checking. If the PR is a bug fix which does not contain major feature change and has been approved before code freeze, an exception is not needed and sig leads could merge it by tagging milestone label.  :)

> With the code freeze in effect, I am not sure we can get the HC changes done. I wanted to break it out so that it didn't drag this main PR down with it :)
> 
> This PR was approved before freeze and is not part of a KEP (limited impact) - do we need an exception for it Cici Huang ?

thockin commented:
@kubernetes/sig-scalability 
@wojtek-t 
@mborsz 

FYI re scale - this should be an improvement for real clusters (minor because etp=Local, but eventually for all services)

alexanderConstantinescu commented:
@thockin : I am concerned the first commit might be wrong, specifically around removing nodes from the LB set based on the schedulability predicate. 

It seems it was removed with #90823 which listed a production outage caused by the fact that we were syncing LBs as a cause by the schedulability constraint changing on a set of nodes.

> This caused a production outage when nodes in an autoscaling group were marked unschedulable so the autoscaler would gracefully terminate them - instead the ingress controllers on those nodes were immediately removed from the LB pool and no new ingress requests could be accepted.

That makes me wonder why we even decided to sync updates to nodes which are experiencing a change to this attribute at all. That was added with PR: https://github.com/kubernetes/kubernetes/pull/90769 - which doesn't list any good reasons for adding it...

So:

- #90823 tells me we should not be excluding nodes from the LB if they are not schedulable 
- #90769 tells me we should (possibly) not even care about the schedulability at all...

Now we did the inverse it seems. Sorry for releasing this now....

alexanderConstantinescu commented:
@thockin : I am concerned the first commit might be wrong, specifically around removing nodes from the LB set based on the schedulability predicate.

It seems it was removed with https://github.com/kubernetes/kubernetes/pull/90823 which listed a production outage caused by the fact that we were syncing LBs as a cause by the schedulability constraint changing on a set of nodes.

> This caused a production outage when nodes in an autoscaling group were marked unschedulable so the autoscaler would gracefully terminate them - instead the ingress controllers on those nodes were immediately removed from the LB pool and no new ingress requests could be accepted.

That makes me wonder why we even decided to sync updates to nodes which are experiencing a change to this attribute at all. That was added with PR: https://github.com/kubernetes/kubernetes/pull/90769 - which doesn't list any good reasons for adding it...

So:

- https://github.com/kubernetes/kubernetes/pull/90823 tells me we should not be excluding nodes from the LB if they are not schedulable
- https://github.com/kubernetes/kubernetes/pull/90769 tells me we should (possibly) not even care about the schedulability at all...

Now we did the inverse it seems. Sorry for releasing this now....

aojea commented on staging/src/k8s.io/cloud-provider/controllers/service/controller.go:
```diff
@@ -772,24 +748,29 @@ func (c *Controller) nodeSyncInternal(ctx context.Context, workers int) {
+		return retSuccess
 	}
-
-	if err := c.lockedUpdateLoadBalancerHosts(svc, hosts); err != nil {
```
this may have been locked some time ago, but I can 't see where is locked now

zhaizeyu commented:
I find that when I restart kubelet, the endpoints of the service disappear for a short time, but the node is not in the not ready state. This problem appeared in v1.22.1 but disappeared in v1.25.3. My service is type=ClusterIP. Is it related to this issue?

**During this period, kube-controller prints only the following logs:**
I0801 06:05:09.544857       9 controller.go:262] Node changes detected, triggering a full node sync on all loadbalancer services
I0801 06:05:09.654068       9 controller.go:262] Node changes detected, triggering a full node sync on all loadbalancer services
I0801 06:05:10.762830       9 utils.go:366] couldn't find ipfamilies for headless service: ns000000000000000000001/fkft7-nslb-north-svc likely because controller manager is likely connected to an old apiserver that does not support ip families yet. The service endpoint slice will use dual stack families until api-server default it correctly
I0801 06:05:34.111447       9 utils.go:366] couldn't find ipfamilies for headless service: ns000000000000000000001/fkft7-nslb-north-svc likely
